### ISSUE TITLE
Azure IoT Hub Connection

### ISSUE THREAD

**Issue by johnvanv** [shamblett/mqtt_client#185]
I hesitate to call this an issue, but I have been trying all day and can't make this work. Can you provide an example of how to connect to an Azure IoT Hub device?

**Comment by shamblett** [comment]
I dont use Azure myself, I use GCP and I can't recall any Azure related issues that say the user couldn't connect. Presumably you have done the obvious checks like setting the client version to be that same as the Azure broker expects. If you turn on logging on the client and post the output from a connect I may be able to help.

**Comment by johnvanv** [comment]
Thanks for your quick response. The mqtt_client library was fine -- the problems were my own. 

The most challenging to fix was due to a mistake in the MS documentation. Their "Communicate with your IoT hub using the MQTT protocol" page states the username should end in "/?api-version=2018-06-30". Eventually, I found on the "Control access to IoT Hub" page that excludes the api-version from the username. 

For reference, I used these properties on MqttServerClient:
- secure = true
- setProtocolV311()
- port = 8883
- useWebSocket = false
- server = {iothubhostname}
- clientIdentifier = {deviceId}
- username = {iothubhostname}/{deviceId}
- password = 'SharedAccessSignature sig={signature-string}&se={expriy}&sr={URL-encoded-resourceURI}'

The password is a SAS token string that I generated using the Azure IoT Tools for Visual Studio Code. The tokens expire so I will need to periodically create new tokens and get them on to the device (that's a problem for another day).

I also added the 'DigiCert Baltimore Root Certificate' to the security context by bundling it as an asset in the startup code:
      rootBundle.load('assets/certs/baltimore.pem')
        .then((data) {
          final context = SecurityContext.defaultContext;
          context.setTrustedCertificatesBytes(data.buffer.asUint8List());
          _client.securityContext = context;
        });
https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-mqtt-support#tlsssl-configuration

Hopefully that's useful if anyone else has Azure IoT challenges.
